tools/unicode_properties_parse/format_width_estimate_intervals.cpp

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// The following code generates data for `_Width_estimate_intervals_v2` in <format>.
+
+#include <charconv>
+#include <cstdint>
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <regex>
+#include <source_location>
+#include <string>
+#include <system_error>
+#include <vector>
+using namespace std;
+
+void verify(bool test, const char* msg, source_location loc = source_location::current()) {
+    if (!test) {
+        cerr << "Error at line " << loc.line() << ": " << msg << endl;
+        exit(EXIT_FAILURE);
+    }
+}
+constexpr const char* impl_assertion_failed = "impl assertion failed";
+
+struct range_u {
+    uint32_t from;
+    uint32_t to;
+    constexpr range_u(uint32_t f, uint32_t t) : from(f), to(t) {}
+    constexpr explicit range_u(uint32_t v) : from(v), to(v) {}
+};
+
+enum class width_u : bool { is_1 = false, is_2 = true };
+
+class table_u {
+private:
+    // A valid Unicode code point won't exceed `max_u`.
+    static constexpr uint32_t max_u = 0x10'ffff;
+    vector<width_u> table;
+
+public:
+    table_u() : table(max_u + 1, width_u::is_1) {}
+
+    void fill_range(const range_u rng, const width_u width) {
+        const auto [from, to] = rng;
+        verify(from <= to, impl_assertion_failed);
+        verify(to <= max_u, impl_assertion_failed);
+        for (uint32_t u = from; u <= to; ++u) {
+            table[u] = width;
+        }
+    }
+
+    void print_intervals() const {
+        // Print table for `_Width_estimate_intervals_v2`.
+        int c        = 0;
+        width_u last = table[0];
+        for (uint32_t u = 0; u <= max_u; ++u) {
+            if (table[u] != last) {
+                cout << "0x" << hex << uppercase << u << "u, ";
+                if (++c == 12) {
+                    c = 0;
+                    cout << endl;
+                }
+            }
+            last = table[u];
+        }
+        cout << endl;
+    }
+
+    void print_clusters_1_vs_2(const table_u& other) const {
+        vector<bool> cluster_table(max_u + 1, false);
+        for (uint32_t u = 0; u <= max_u; ++u) {
+            if (table[u] == width_u::is_1 && other.table[u] == width_u::is_2) {
+                cluster_table[u] = true;
+            }
+        }
+
+        for (uint32_t u = 0; u <= max_u; ++u) {
+            if (cluster_table[u]) {
+                const uint32_t from = u;
+                uint32_t to         = from;
+                while (to + 1 <= max_u && cluster_table[to + 1]) {
+                    ++to;
+                }
+                if (from == to) {
+                    cout << hex << uppercase << "U+" << from << endl;
+                } else {
+                    cout << hex << uppercase << "U+" << from << "..U+" << to << endl;
+                }
+                u = to;
+            }
+        }
+    }
+};
+
+table_u get_table_cpp20() {

```

Comment:
Is `_cpp20/cpp23` good names(also used in some variables)? (talked about in this thread https://github.com/microsoft/STL/pull/3903/files#r1272896352)

Reply:
It's fine.